File: 兴趣班入门百练讲义/NQ100/NQ076/NQ076.cpp
#include <iostream>
#include <algorithm>
using namespace std;
typedef long long LL;
int qmi(int a, int k, int p)
{
    int res = 1; //初始值为1
    while (k)
    {
        if (k & 1)                 //b的最末尾为1
            res = (LL)res * a % p; //必须用long long防止溢出
        k >>= 1;                   //去掉b最低位
        a = (LL)a * a % p;
    }
    return res;
}

int main()
{
    int q;
    scanf("%d", &q); //数据很大，需要用LL读入
    while (q--)
    {
        int a, k, p;
        scanf("%d%d%d", &a, &k, &p); //读入a b p

        printf("%d\n", qmi(a, k, p)); //计算快速幂
    }
    return 0;
}

File: 兴趣班入门百练讲义/NQ100/NQ076/NQ076.py
# NQ076 二进制取幂法 Python 3.5 版本

def qmi(a, k, p):
    res = 1  # 初始值为1
    while k:
        if k & 1:  # k的最末尾为1
            res = res * a % p  # 计算并取模
        k >>= 1  # 去掉k最低位
        a = a * a % p  # 提前计算a的平方并取模
    return res

if __name__ == "__main__":
    q = int(input().strip())  # 读入q，表示有q组数据
    for _ in range(q):
        a, k, p = map(int, input().strip().split())  # 读入a, k, p
        print(qmi(a, k, p))  # 计算快速幂并输出结果